src/System.Runtime.Loader/tests/CollectibleAssemblyLoadContextTest.cs

```diff
@@ -0,0 +1,554 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Xunit;
+using System.Threading;
+
+namespace System.Runtime.Loader.Tests
+{
+
+    public partial class AssemblyLoadContextTest
+    {
+        // Tests related to Collectible assemblies
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void CreateAndLoadContext(CollectibleChecker checker)
+        {
+            var alc = new ResourceAssemblyLoadContext(true);
+            checker.SetAssemblyLoadContext(0, alc);
+
+            alc.Unload();
+
+            // Check that any attempt to load an assembly after an explicit Unload will fail
+            Assert.Throws<InvalidOperationException>(() => alc.LoadFromAssemblyPath("none.dll"));
+
+            alc = null;
+        }
+
+        [Fact]
+        public static void Unload_CollectibleWithNoAssemblyLoaded()
+        {
+            // Use a collectible ALC + Unload
+            // Check that we receive the Unloading event
+
+            var checker = new CollectibleChecker(1);
+            CreateAndLoadContext(checker);
+            checker.GcAndCheck();
+        }
+
+        [Fact]
+        public static void Finalizer_CollectibleWithNoAssemblyLoaded()
+        {
+            // Use a collectible ALC, let the finalizer call Unload
+            // Check that we receive the Unloading event
+
+            var checker = new CollectibleChecker(1);
+            // Create the ALC in another method to allow the finalizer to run
+            var weakRef = CreateCollectible(checker);
```

Comment:
The CoreFX coding convention frowns on using `var` in this context. https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md